### DISCUSSION TITLE
EMQX behavior on retries without disconnection

### DISCUSSION THREAD

**Discussion by shivamkm07** [emqx/emqx#9447] [Q&A]
EMQX docs state that it is `Fully MQTT 5.0: 100% compliant with MQTT 5.0 and 3.x standard`. However, there are some differences in mqtt3 and mqtt5 specifications. The one I am concerned here is about retries:
Currently emqx seems to be retrying on unacked messages with QOS 1, after a fixed timeout (Tested with v5.0.11). This is fine with MQTT 3 but not with MQTT 5. MQTT5 clearly states that "Clients and Servers MUST NOT resend messages at any other time other than reconnection". So is it being tracked to be removed from emqx in future releases?

Please note that we (https://github.com/dapr/) are using MQTT as Pubsub component (https://docs.dapr.io/reference/components-reference/supported-pubsub/setup-mqtt/) with EMQX as broker. Dapr provides at least once guarantee for messages published with qos > 1. So if emqx decides to deprecate retrying for unacked messages as per MQTT5,  it would violate our guarantee and in the worst case we might need to deprecate MQTT component.

**Comment by zmstone** [comment]
Hi @shivamkm07 

We are considering to deprecate the `retry_interval` in the coming v5 releases. (no change for EMQX 4.x).

I cannot seem to find in the spec where it allows periodic redelivery even in [v3.x spec](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html).

Instead, we found this:
> 4.4 Message delivery retry
When a Client reconnects with CleanSession set to 0, both the Client and Server MUST re-send any unacknowledged PUBLISH Packets (where QoS > 0) and PUBREL Packets using their original Packet Identifiers [MQTT-4.4.0-1]. This is the only circumstance where a Client or Server is REQUIRED to redeliver messages.

Mind the "the only circumstance" part.
We'd like to know why, without a periodic retry, EMQX would fail dapr's delivery guarantee?

**Reply by shivamkm07** [reply]
So when you say `retry_interval` would be deprecated, does that mean periodic redelivery would be deprecated as well? Or just the option would be non-configurable by user and EMQX would be internally using a default set value? Currently as per emqx docs, `retry-interval` is said to be deprecated since 5.0.4(https://www.emqx.io/docs/en/v5/admin/cfg.html#authz-http-get) but on testing I found emqx still does periodic retries on failures. It would help if you could just clarify what is in consideration.

Also as you correctly pointed out, mqtt 3.x spec says `This is the only circumstance where a Client or Server is REQUIRED to redeliver messages.` So it does not prohibit re-delivery. It's just that is not required as per the specification. On the other hand, mqtt v5 also adds`Clients and Servers MUST NOT resend messages at any other time` ( https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901238). 

As for Dapr, we rely on broker re-delivery of messages to ensure messages are received to the subscriber at least once in case of failures. So if periodic retry is deprecated, broker won't resend messages in case of failures (unless there is disconnection and reconnection from client, which is not a regular occurrence in microservices) and so Dapr's at least once delivery of messages would be deprecated. Note here that In Dapr, there is a Dapr sidecar which runs alongside application container and it is the dapr sidecar which subscribes to the emqx broker. So if application crashes, there is no disconnection and reconnection and hence messages are not retried.
